### PR TITLE
GSASIIRefineFitPeaks enable Pawley refinement

### LINKED ISSUE #22243: GSASIIRefineFitPeaks Pawley refinement needs support
A refinement recipe for Pawley refinement from GSASIIRefineFitPeaks needs to be devised and implemented in the algorithm, as currently only Rietveld works properly

### PR DESCRIPTION
**Pull request by josephframsay**
GSASIIRefineFitPeaks now supports Pawley refinement. In addition, the system tests for the algorithm have been made more stringent, and the usage examples in the documentation (which are not run on the build server) were brought up to date with the current version of the algorithm

**To test:**

Run the GSASIIRefineFitPeaks system tests. **Note these are not currently run on the build server, so please validate that they pass on your machine**. You'll need GSASII installed - there are instructions for doing this [here](http://docs.mantidproject.org/nightly/algorithms/GSASIIRefineFitPeaks-v1.html#id8).

To do a refinement, use the test data [here](https://github.com/mantidproject/mantid/files/1739753/GSASIIRefineFitPeaks.zip)

```
import os

# Replace this with your gsas-ii installation directory
path_to_gsas = "/directory/containing/GSASIIscriptable.py"

# Replace this with a location to save the GSAS-II project file
gsas_proj_file = "/somewhere/local/Test.gpx"

data_dir = "/wherever/you/extracted/test/data"
input_file = lambda file: os.path.join(data_dir, file)

phase_files = ",".join([input_file("Fe-{}.cif".format(phase)) for phase in ["alpha", "gamma", "epsilon"]])
iparams_file = input_file("template_ENGINX_241391_236516_North_bank.prm")
input_ws = Load(Filename=input_file("ENGINX_280625_focused_bank_1.nxs"))

fitted_peaks, lattice_params, rwp, sigma, gamma = \
    GSASIIRefineFitPeaks(InputWorkspace=input_ws,
                         PhaseInfoFiles=phase_files,
                         InstrumentFile=iparams_file,
                         RefinementMethod="Pawley refinement",
                         SaveGSASIIProjectFile=gsas_proj_file,
                         PathToGSASII=path_to_gsas,
                         RefineSigma=True,
                         RefineGamma=True,
                         MuteGSASII=True) # Take this out if you're interested in GSAS-II output
```

Make sure the fit looks good, and nothing breaks unexpectedly. 

Fixes #22243 

**Release Notes** 
[Here](https://github.com/mantidproject/mantid/compare/22243_gsasiirefinefitpeaks_pawley#diff-b2bd2540c595f643521322a749fbd5dc)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by soininen** [comment]
@josephframsay Please resolve the merge conflicts.

**Comment by soininen** [comment]
The two system tests do not pass. I get the following output:
```
Tue, 24 Apr 2018 08:36:51: Executing GSASIIRefineFitPeaksTest.GSASIIRefineFitPeaksPawleyTest
[...]
/users/soininen/g2conda/GSASII/GSASIIstrIO.py:2794: RuntimeWarning: invalid value encountered in double_scalars
  wtFr = hapData['Scale'][0]*General['Mass']/wtFrSum[hId]
GSASIIRefineFitPeaks-[Notice] GSASIIRefineFitPeaks successful, Duration 5.36 seconds
RESULT|iteration time_taken|1 5.73
Traceback (most recent call last):
  File "/tmp/tmpMcWsUO", line 9, in <module>
    exitcode = systest.returnValidationCode(99)
  File "/home/cs/soininen/Sources/mantid/Testing/SystemTests/lib/systemtests/stresstesting.py", line 375, in returnValidationCode
    if self.doValidation():
  File "/home/cs/soininen/Sources/mantid/Testing/SystemTests/lib/systemtests/stresstesting.py", line 345, in doValidation
    validation = self.validate()
  File "/home/cs/soininen/Sources/mantid/Testing/SystemTests/tests/analysis/GSASIIRefineFitPeaksTest.py", line 109, in validate
    self.assertAlmostEqual(self.rwp, self._get_expected_rwp(), delta=1e-6)
  File "/usr/lib/python2.7/unittest/case.py", line 561, in assertAlmostEqual
    raise self.failureException(msg)
AssertionError: 24.05162012368024 != 24.051622 within 1e-06 delta
```

The above looks like a rounding error.

The second test just says the reference file is different... I wonder what went wrong?

```
Tue, 24 Apr 2018 08:37:01: Executing GSASIIRefineFitPeaksTest.GSASIIRefineFitPeaksRietveldTest
Executing test script '/users/soininen/mantid-build-Default/bin/mantidpython --classic /tmp/tmp52lMYW'
ConfigService-[Information] Unable to locate directory at: /etc/mantid/instrument
ConfigService-[Information] This is Mantid version 3.12.20180423.803 revision g86a3604
ConfigService-[Information] running on johnsonpc3 starting 2018-04-24T06:37Z
ConfigService-[Information] Properties file(s) loaded: /users/soininen/mantid-build-Default/bin/Mantid.properties, /home/cs/soininen/.mantid/Mantid.user.properties
ConfigService-[Information] Logging to: /home/cs/soininen/.mantid/mantid.log
ConfigService-[Information] Unable to locate directory at: /etc/mantid/instrument
FrameworkManager-[Notice] Welcome to Mantid 3.12.20180423.803
FrameworkManager-[Notice] Please cite: http://dx.doi.org/10.1016/j.nima.2014.07.029 and this release: http://dx.doi.org/10.5286/Software/Mantid
FrameworkManager-[Information] Instrument updates disabled - cannot update instrument definitions.
FrameworkManager-[Information] Version check disabled.
/usr/lib/python2.7/dist-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
Load-[Notice] Load started
Load-[Information] Algorithm: Load v1
Load-[Information] Execution Date: 2018-Apr-24 06:37:05
Load-[Information] Execution Duration: -1 seconds
Load-[Information] Parameters:
Load-[Information]   Name: Filename, Value: /users/soininen/m ... NGINX00256663.nxs, Default?: No, Direction: Input
Load-[Information]   Name: OutputWorkspace, Value: input_ws, Default?: No, Direction: Output
Load-[Information]   Name: LoaderName, Value: LoadNexusProcessed, Default?: No, Direction: Output
Load-[Information]   Name: LoaderVersion, Value: 1, Default?: No, Direction: Output
Load-[Information]   Name: SpectrumMin, Value: 1, Default?: Yes, Direction: Input
Load-[Information]   Name: SpectrumMax, Value: 2147483647, Default?: Yes, Direction: Input
Load-[Information]   Name: SpectrumList, Value: , Default?: Yes, Direction: Input
Load-[Information]   Name: EntryNumber, Value: 0, Default?: Yes, Direction: Input
Load-[Information]   Name: LoadHistory, Value: 1, Default?: Yes, Direction: Input
Load-[Information]   Name: FastMultiPeriod, Value: 1, Default?: Yes, Direction: Input
Load-[Information] Using LoadNexusProcessed version 1.
InstrumentDefinitionParser-[Information] Loading geometry cache from /home/cs/soininen/.mantid/instrument/geometryCache/ENGIN-X9af32bc1eb2fa7edd195ae0abce9fa020887b0f2.vtp
Load-[Notice] Load successful, Duration 0.10 seconds
GSASIIRefineFitPeaks-[Notice] GSASIIRefineFitPeaks started
GSASIIRefineFitPeaks-[Information] Algorithm: GSASIIRefineFitPeaks v1
GSASIIRefineFitPeaks-[Information] Execution Date: 2018-Apr-24 06:37:05
GSASIIRefineFitPeaks-[Information] Execution Duration: -1 seconds
GSASIIRefineFitPeaks-[Information] Parameters:
GSASIIRefineFitPeaks-[Information]   Name: RefinementMethod, Value: Rietveld refinement, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: InputWorkspace, Value: input_ws, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: WorkspaceIndex, Value: 0, Default?: Yes, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: InstrumentFile, Value: /users/soininen/m ... 16_North_bank.prm, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: PhaseInfoFiles, Value: /users/soininen/m ... Test/Fe-alpha.cif, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: PathToGSASII, Value: /users/soininen/g2conda/GSASII, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: XMin, Value: 10000, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: XMax, Value: 40000, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: RefineSigma, Value: 1, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: RefineGamma, Value: 1, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: OutputWorkspace, Value: FittedPeaks, Default?: No, Direction: Output
GSASIIRefineFitPeaks-[Information]   Name: LatticeParameters, Value: LatticeParameters, Default?: Yes, Direction: Output
GSASIIRefineFitPeaks-[Information]   Name: Rwp, Value: 0, Default?: Yes, Direction: Output
GSASIIRefineFitPeaks-[Information]   Name: Sigma, Value: 0, Default?: Yes, Direction: Output
GSASIIRefineFitPeaks-[Information]   Name: Gamma, Value: 0, Default?: Yes, Direction: Output
GSASIIRefineFitPeaks-[Information]   Name: SaveGSASIIProjectFile, Value: /users/soininen/m ... sRietveldTest.gpx, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: PawleyDMin, Value: 1, Default?: Yes, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: PawleyNegativeWeight, Value: 0, Default?: Yes, Direction: Input
GSASIIRefineFitPeaks-[Information]   Name: MuteGSASII, Value: 1, Default?: No, Direction: Input
GSASIIRefineFitPeaks-[Information] Suppressing stdout
SaveFocusedXYE-[Notice] SaveFocusedXYE started (child)
SaveFocusedXYE-[Information] Algorithm: SaveFocusedXYE v1
SaveFocusedXYE-[Information] Execution Date: 2018-Apr-24 06:37:06
SaveFocusedXYE-[Information] Execution Duration: -1 seconds
SaveFocusedXYE-[Information] Parameters:
SaveFocusedXYE-[Information]   Name: InputWorkspace, Value: __TMP0x5b5e780, Default?: No, Direction: Input
SaveFocusedXYE-[Information]   Name: Filename, Value: /tmp/GSASIIRefine ... d_spectrum-0.fxye, Default?: No, Direction: Input
SaveFocusedXYE-[Information]   Name: SplitFiles, Value: 0, Default?: No, Direction: Input
SaveFocusedXYE-[Information]   Name: StartAtBankNumber, Value: 0, Default?: Yes, Direction: Input
SaveFocusedXYE-[Information]   Name: Append, Value: 0, Default?: Yes, Direction: Input
SaveFocusedXYE-[Information]   Name: IncludeHeader, Value: 0, Default?: No, Direction: Input
SaveFocusedXYE-[Information]   Name: Format, Value: XYE, Default?: Yes, Direction: Input
SaveFocusedXYE-[Notice] SaveFocusedXYE successful, Duration 0.04 seconds
GSASIIRefineFitPeaks-[Notice] GSASIIRefineFitPeaks successful, Duration 5.71 seconds
RESULT|iteration time_taken|1 5.85
Load-[Notice] Load started
Load-[Information] Algorithm: Load v1
Load-[Information] Execution Date: 2018-Apr-24 06:37:11
Load-[Information] Execution Duration: -1 seconds
Load-[Information] Parameters:
Load-[Information]   Name: Filename, Value: /users/soininen/m ... veldFitParams.nxs, Default?: No, Direction: Input
Load-[Information]   Name: OutputWorkspace, Value: RefFile, Default?: No, Direction: Output
Load-[Information]   Name: LoaderName, Value: LoadNexusProcessed, Default?: No, Direction: Output
Load-[Information]   Name: LoaderVersion, Value: 1, Default?: No, Direction: Output
Load-[Information]   Name: SpectrumMin, Value: 1, Default?: Yes, Direction: Input
Load-[Information]   Name: SpectrumMax, Value: 2147483647, Default?: Yes, Direction: Input
Load-[Information]   Name: SpectrumList, Value: , Default?: Yes, Direction: Input
Load-[Information]   Name: EntryNumber, Value: 0, Default?: Yes, Direction: Input
Load-[Information]   Name: LoadHistory, Value: 1, Default?: Yes, Direction: Input
Load-[Information]   Name: FastMultiPeriod, Value: 1, Default?: Yes, Direction: Input
Load-[Information] Using LoadNexusProcessed version 1.
Load-[Notice] Load successful, Duration 0.02 seconds
CompareWorkspaces-[Notice] CompareWorkspaces started
CompareWorkspaces-[Information] Algorithm: CompareWorkspaces v1
CompareWorkspaces-[Information] Execution Date: 2018-Apr-24 06:37:11
CompareWorkspaces-[Information] Execution Duration: -1 seconds
CompareWorkspaces-[Information] Parameters:
CompareWorkspaces-[Information]   Name: Workspace1, Value: LatticeParameters, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: Workspace2, Value: RefFile, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: Tolerance, Value: 1e-08, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: CheckType, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckAxes, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckSpectraMap, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckInstrument, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckMasking, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckSample, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: ToleranceRelErr, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckAllData, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: NumberMismatchedSpectraToPrint, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: DetailedPrintIndex, Value: 2147483647, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: Result, Value: 0, Default?: Yes, Direction: Output
CompareWorkspaces-[Information]   Name: Messages, Value: compare_msgs, Default?: Yes, Direction: Output
CompareWorkspaces-[Notice] The workspaces "LatticeParameters" and "RefFile" matched!
CompareWorkspaces-[Notice] CompareWorkspaces successful, Duration 0.00 seconds
Load-[Notice] Load started
Load-[Information] Algorithm: Load v1
Load-[Information] Execution Date: 2018-Apr-24 06:37:11
Load-[Information] Execution Duration: -1 seconds
Load-[Information] Parameters:
Load-[Information]   Name: Filename, Value: /users/soininen/m ... ldFittedPeaks.nxs, Default?: No, Direction: Input
Load-[Information]   Name: OutputWorkspace, Value: RefFile, Default?: No, Direction: Output
Load-[Information]   Name: LoaderName, Value: LoadNexusProcessed, Default?: No, Direction: Output
Load-[Information]   Name: LoaderVersion, Value: 1, Default?: No, Direction: Output
Load-[Information]   Name: SpectrumMin, Value: 1, Default?: Yes, Direction: Input
Load-[Information]   Name: SpectrumMax, Value: 2147483647, Default?: Yes, Direction: Input
Load-[Information]   Name: SpectrumList, Value: , Default?: Yes, Direction: Input
Load-[Information]   Name: EntryNumber, Value: 0, Default?: Yes, Direction: Input
Load-[Information]   Name: LoadHistory, Value: 1, Default?: Yes, Direction: Input
Load-[Information]   Name: FastMultiPeriod, Value: 1, Default?: Yes, Direction: Input
Load-[Information] Using LoadNexusProcessed version 1.
Load-[Notice] Load successful, Duration 0.11 seconds
CompareWorkspaces-[Notice] CompareWorkspaces started
CompareWorkspaces-[Information] Algorithm: CompareWorkspaces v1
CompareWorkspaces-[Information] Execution Date: 2018-Apr-24 06:37:11
CompareWorkspaces-[Information] Execution Duration: -1 seconds
CompareWorkspaces-[Information] Parameters:
CompareWorkspaces-[Information]   Name: Workspace1, Value: FittedPeaks, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: Workspace2, Value: RefFile, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: Tolerance, Value: 1e-08, Default?: No, Direction: Input
CompareWorkspaces-[Information]   Name: CheckType, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckAxes, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckSpectraMap, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckInstrument, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckMasking, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckSample, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: ToleranceRelErr, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: CheckAllData, Value: 0, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: NumberMismatchedSpectraToPrint, Value: 1, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: DetailedPrintIndex, Value: 2147483647, Default?: Yes, Direction: Input
CompareWorkspaces-[Information]   Name: Result, Value: 0, Default?: Yes, Direction: Output
CompareWorkspaces-[Information]   Name: Messages, Value: compare_msgs, Default?: Yes, Direction: Output
CompareWorkspaces-[Notice] The workspaces did not match: Data mismatch
CompareWorkspaces-[Notice] CompareWorkspaces successful, Duration 0.07 seconds
SaveNexus-[Notice] SaveNexus started
SaveNexus-[Information] Algorithm: SaveNexus v1
SaveNexus-[Information] Execution Date: 2018-Apr-24 06:37:11
SaveNexus-[Information] Execution Duration: -1 seconds
SaveNexus-[Information] Parameters:
SaveNexus-[Information]   Name: InputWorkspace, Value: FittedPeaks, Default?: No, Direction: Input
SaveNexus-[Information]   Name: Filename, Value: /users/soininen/m ... eaks-mismatch.nxs, Default?: No, Direction: Input
SaveNexus-[Information]   Name: Title, Value: , Default?: Yes, Direction: Input
SaveNexus-[Information]   Name: WorkspaceIndexMin, Value: 0, Default?: Yes, Direction: Input
SaveNexus-[Information]   Name: WorkspaceIndexMax, Value: 2147483647, Default?: Yes, Direction: Input
SaveNexus-[Information]   Name: WorkspaceIndexList, Value: , Default?: Yes, Direction: Input
SaveNexus-[Information]   Name: Append, Value: 0, Default?: Yes, Direction: Input
SaveNexus-[Notice] SaveNexus successful, Duration 0.07 seconds
GSASIIRefineFitPeaksRietveldTest
Workspace FittedPeaks not equal to its reference file
RESULT|memory footprint increase|56.66796875
```

Let me know if you need more information to investigate the system tests. These were run under Ubuntu 16.04.

**Comment by josephframsay** [comment]
@soininen I get a failure on Ubuntu as well, but a pass on Windows. I think this is to do with different versions of GSAS-II (or possibly GSAS-II behaving differently on different OS), so the only thing I can think of is to turn the tolerance up a little bit

File: Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py [hunk 1]
```diff
@@ -113,7 +115,9 @@ def PyInit(self):
 
         self.declareProperty(name=self.PROP_PAWLEY_DMIN, defaultValue=1.0, direction=Direction.Input,
                              doc="For Pawley refiment: as defined in GSAS-II, the minimum d-spacing to be used in a "
-                                 "Pawley refinement. Please refer to the GSAS-II documentation for full details.")
+                                 "Pawley refinement. Please refer to the GSAS-II documentation for full details. Note, "
+                                 "if this corresponds to a TOF value less than {} or the lowest TOF value in the data, "
+                                 "the greatest of the 3 values is used".format(self.PROP_XMIN))
```
Comment (soininen) [inline Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py]: If adding `PawleyDMin=1000` to the test script, one gets a bit messy error `RuntimeError: index -1 is out of bounds for axis 0 with size 0`. Would it be possible to add some validation to this parameter to get more meaningful error messages?

Comment (soininen) [reply Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py]: Another question: should the doc read `if this corresponds to a TOF value less than self.PROP_XMIN or the lowest TOF value in the data the greatest of the 3 values is used as self.PROP_XMIN`?

**Review by soininen — CHANGES_REQUESTED** [review]
The code changes look generally very good. There is just two minor issues.

**Comment by soininen** [comment]
Tested this with the provided data set and Python script. Tried to play with the algorithm's properties a bit and usually got a meaningful error message.

The fit looks as following:
pawley_refinement

I'm not sure if the fit 'looks good'; the low TOF values are not fitted, and the peak on the left side of the two Mount Everests is not fitted at all. Is this expected?

**Comment by josephframsay** [comment]
@soininen The low TOF values are not fitted because the lower limit in this case is the default PawleyDMin, which corresponds to a TOF of about 18000. If you lower PawleyDMin you'll see more of the low TOF peaks come into the red line.

Regarding the neighbouring peak (Nuptse?), the phase files I've used are for iron but the sample was steel (apologies, a bit of an oversight). If you go and download the [cif file for carbon](http://www.crystallography.net/cod/9014004.html) and include it in the list of phase files, you'll see this peak gets fitted

**Review by soininen — APPROVED** [review]

**Pull request merged**